chorus_lib/src/transport/local.rs

```diff
@@ -5,52 +5,107 @@ use std::sync::Arc;
 
 use serde_json;
 
-use crate::core::{Portable, Transport};
+use std::marker::PhantomData;
+
+use crate::LocationSet;
+
+use crate::core::{ChoreographyLocation, HCons, HList, Portable, Transport};
 use crate::utils::queue::BlockingQueue;
 
 type QueueMap = HashMap<String, HashMap<String, BlockingQueue<String>>>;
 
-/// The local transport.
-///
-/// This transport uses a blocking queue to allow for communication between threads. Each location must be executed in its thread.
-///
-/// Unlike network-based transports, all locations must share the same `LocalTransport` instance. The struct implements `Clone` so that it can be shared across threads.
-#[derive(Clone)]
-pub struct LocalTransport {
-    internal_locations: Vec<String>,
+/// A Transport channel used between multiple `Transport`s.
+pub struct LocalTransportChannel<L: HList> {
+    /// The location set where the channel is defined on.
+    location_set: std::marker::PhantomData<L>,
     queue_map: Arc<QueueMap>,
 }
 
-impl LocalTransport {
-    /// Creates a new `LocalTransport` instance from a list of locations.
-    pub fn from(locations: &[&str]) -> Self {
+impl<L: HList> Clone for LocalTransportChannel<L> {
+    fn clone(&self) -> Self {
+        LocalTransportChannel {
+            location_set: PhantomData,
+            queue_map: self.queue_map.clone(),
+        }
+    }
+}
+
+impl LocalTransportChannel<LocationSet!()> {
+    /// Creates a new `LocalTransportChannel` instance
+    pub fn new() -> Self {
+        Self {
+            location_set: PhantomData,
+            queue_map: HashMap::new().into(),
+        }
+    }
+}
+
+impl<L: HList> LocalTransportChannel<L> {
+    /// Adds a new location to the set of locations in the `LocalTransportChannel`.
+    pub fn with<NewLocation: ChoreographyLocation>(
+        self,
+        _location: NewLocation,
+    ) -> LocalTransportChannel<HCons<NewLocation, L>> {
         let mut queue_map: QueueMap = HashMap::new();
```

Comment:
I know it's not a big deal, but you can probably take `mut self` and reuse the hashmap without creating a new one all over every time we call `with`

Reply:
queue_map is always an `Arc`, and if you want it to be muted, you have to change the type to something else, I think. And even if you are able to do that, the code itself will become a bit harder to understand (Here it's clear what queue_map contains if someone reads the code, but if you want it to be changed so that only new information is added to it, someone might need to read it a couple of times before they understand all the content, not just how it's updated every time)

Reply:
This is a library and that complexity won't be exposed to the user. IMO that readability issue is not a valid explanation for creating a nested hash map.